File: 2023/day05/go/seed_map/seed_map.go
package seedmap

import (
	"errors"
)

type SeedMap struct {
	source uint64
	target uint64
	size   uint64
}

func New(source uint64, target uint64, size uint64) SeedMap {
	seedMap := SeedMap{source: source, target: target, size: size}
	return seedMap
}

// input string is typically "1 2 5\n"
func NewFromString(inputString string) (SeedMap, error) {
	return SeedMap{}, errors.New("invalid seed map string")
}

func (sm SeedMap) MapSeed(seed_value *uint64) bool {
	if *seed_value < sm.source ||
		*seed_value >= (sm.source+sm.size) {
		return false
	}

	*seed_value = (*seed_value - sm.source) + sm.target
	return true
}

File: 2023/day05/experiments/python_rewrite/seed_map/seed_map.py



class SeedMap:

    def __init__(self, source, target, size):
        self.source = source
        self.source_end = source + size
        self.target = target
        self.target_end = target + size
        self.size = size

    def init_from_string(input_string: str):
        number_strings = input_string.split(" ")
        return SeedMap(int(number_strings[1]), int(number_strings[0]), int(number_strings[2]))

    def map_seed(self, value:int):
        if(value >= self.source) and (value < self.source_end):
            return (value - self.source) + self.target
        return None

    def map_seed_inverse(self, value:int):
        if(value >= self.target) and (value < self.target_end):
            return value - self.target + self.source